src/rgw/rgw_dedup_store.cc

```diff
@@ -0,0 +1,847 @@
+    }
+    else {
+      p_stats->invalid_sha256++;
+      memset(p_rec->s.sha256, 0, sizeof(p_rec->s.sha256));
+    }
+
+    itr = attrs.find(RGW_ATTR_SHARE_MANIFEST);
+    if (itr != attrs.end()) {
+      uint64_t hash = 0;
+      try {
+        auto bl_iter = itr->second.cbegin();
+        ceph::decode(hash, bl_iter);
+        p_rec->s.shared_manifest = hash;
+      } catch (buffer::error& err) {
+        ldpp_dout(dpp, 1) << __func__ << "::ERROR: bad shared_manifest" << dendl;
+        return -1;
+      }
+      ldpp_dout(dpp, 20) << __func__ << "::p_rec->s.shared_manifest=0x" << std::hex
+                         << p_rec->s.shared_manifest << std::dec << dendl;
+      p_rec->s.flags.set_shared_manifest();
+    }
+    else {
+      memset(&p_rec->s.shared_manifest, 0, sizeof(p_rec->s.shared_manifest));
+    }
+
+    return 0;
+  }
+
+  //---------------------------------------------------------------------------
+  static int print_manifest(const DoutPrefixProvider *dpp,
+                            RGWRados                 *rados,
+                            const bufferlist         &manifest_bl)
+  {
+    RGWObjManifest manifest;
+    try {
+      auto bl_iter = manifest_bl.cbegin();
+      decode(manifest, bl_iter);
+    } catch (buffer::error& err) {
+      ldpp_dout(dpp, 1)  << __func__ << "::ERROR: unable to decode manifest" << dendl;
+      return -1;
+    }
+
+    unsigned idx = 0;
+    for (auto p = manifest.obj_begin(dpp); p != manifest.obj_end(dpp); ++p, ++idx) {
+      rgw_raw_obj raw_obj = p.get_location().get_raw_obj(rados);
+      ldpp_dout(dpp, 20) << idx << "] " << raw_obj.oid << dendl;
+    }
+    ldpp_dout(dpp, 20) << "==============================================" << dendl;
+    return 0;
+  }
+
+  //---------------------------------------------------------------------------
+  [[maybe_unused]] static void display_record(const DoutPrefixProvider *dpp,
+                                              RGWRados                 *rados,
+                                              const disk_record_t      &rec)
+  {
+    char buff[DISK_BLOCK_SIZE];
+    ldpp_dout(dpp, 20)  << __func__ << "::Disk Record:\n" << rec << dendl;
+    print_manifest(dpp, rados, rec.manifest_bl);
+
+    ldpp_dout(dpp, 20)  << __func__ << "::Disk Record length = " << rec.length() << dendl;
+    int n = rec.serialize(buff);
+    ldpp_dout(dpp, 20)  << __func__ << "::Disk Record serialized length = " << n << dendl;
+    disk_record_t rec2(buff);
+    ldpp_dout(dpp, 20)  << __func__ << "::Disk Record2:\n" << rec2 << dendl;
+    print_manifest(dpp, rados, rec2.manifest_bl);
+    ldpp_dout(dpp, 20)  << __func__ << "::Disk Record length = " << rec2.length() << dendl;
+    n = rec2.serialize(buff);
+    ldpp_dout(dpp, 20)  << __func__ << "::Disk Record serialized length = " << n << dendl;
+
+    //if (std::memcmp(&rec, &rec2, rec2.length()) == 0) {}
+    if (rec == rec2) {
+      ldpp_dout(dpp, 20)  << __func__ << "::Disk Records match!" << dendl;
+    }
```

Comment:
Suggest removing ! throughout these strings.
These disk records, are they independent of min_alloc_size?  I want to be sure that we don't assuming that min_alloc_size is 4KB.